### PR TITLE
[IMP] evaluation: functions can handle errors

### PR DESCRIPTION
**Pull request by laa-odoo**
With those commits, functions can handle errors more easily. See individual commits for details.

The purpose of this commit is to allow the formulas
to return errors as results.

Functionally we want formulas like TRANSPOSE to be
able to transpose an error without invalidating the
entire evaluation process.

We also want formulas like LOOKUP not to be interrupted
when the search range has one or more errors.

Task: 2969401

### CODE REVIEW

File: src/functions/index.ts [hunk 1]
```diff
@@ -107,6 +108,40 @@ function createComputeFunctionFromDescription(
...
+  if (e instanceof EvaluationError) {
+    e.message = e.message.replace("[[FUNCTION_NAME]]", functionName);
+    return { value: e };
+  }
+  // TODO check this
+  const error = new EvaluationError(
+    e.message.replace("[[FUNCTION_NAME]]", functionName),
+    e?.errorType,
+    e.logLevel !== undefined ? e.logLevel : CellErrorLevel.error
+  );
```
Comment (LucasLefevre) [inline src/functions/index.ts:141]: can be simplified. `e?.errorType` and `e.logLevel` are always undefined because we know `e` is not an `EvaluationError`
```ts
const error = new EvaluationError(e.message.replace("[[FUNCTION_NAME]]", functionName));
```

File: src/functions/module_info.ts [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
-  compute: function (value: () => Maybe<CellValue>): boolean {
-    try {
-      value();
-      return false;
-    } catch (e) {
-      return e?.errorType != CellErrorType.NotAvailable;
-    }
+  compute: function (value: Maybe<CellValue>): boolean {
+    const isErr = value instanceof EvaluationError;
+    const isna = value instanceof NotAvailableError;
```
Comment (LucasLefevre) [inline src/functions/module_info.ts:15]: `isNa`

File: tests/test_helpers/helpers.ts [hunk 3]
```diff
@@ -638,10 +639,13 @@ export function getCellsObject(model: Model, sheetId: UID): Record<string, CellO
   const cells: Record<string, CellObject> = {};
   for (const cell of Object.values(model.getters.getCells(sheetId))) {
     const { col, row } = model.getters.getCellPosition(cell.id);
+
+    const evaluatedCell = model.getters.getEvaluatedCell({ sheetId, col, row });
```
Comment (LucasLefevre) [inline tests/test_helpers/helpers.ts:643]: revert this useless diff

File: src/plugins/ui_core_views/cell_evaluation/evaluation_plugin.ts [hunk 4]
```diff
@@ -219,7 +220,13 @@ export class EvaluationPlugin extends UIPlugin {
     }
     const array = compiledFormula.execute(ranges, ...this.compilationParams);
     if (isMatrix(array)) {
-      return array.map((col) => col.map((row) => row.value));
+      throw new Error("results array are not supported outside the evaluation grid");
+    }
+    if (array.value instanceof EvaluationError) {
+      return array.value.errorType;
+    }
```
Comment (hokolomopo) [inline src/plugins/ui_core_views/cell_evaluation/evaluation_plugin.ts:227]: Seems strange to only here treat error as strings, and not at any other part of the evaluation process.

Also if we want to do that we should test instanceof `Error` (or improve the wrapping of errors into EvaluationError) , because for example WrongNumberOfArgs will return an Error and not an EvaluationError

File: tests/evaluation/evaluation.test.ts [hunk 5]
```diff
@@ -165,17 +165,19 @@ describe("evaluateCells", () => {
   test("error in some function calls", () => {
     const model = new Model();
     setCellContent(model, "A1", '=Sum("asdf")');
-    expect(getEvaluatedCell(model, "A1").value).toBe("#ERROR");
-    expect(getCellError(model, "A1")).toBe(
```
Comment (hokolomopo) [inline tests/evaluation/evaluation.test.ts:169]: why these change? That's a lot of diff and forward port breaking when nothing really had to change, no ? `getEvaluatedCell(model, "A1").value` didn't change, and I'm not sure there is a reason to change `getCellError`

File: src/plugins/ui_core_views/cell_evaluation/evaluator.ts [hunk 6]
```diff
@@ -210,17 +216,21 @@ export class Evaluator {
     }
 
     const cellId = cell.id;
-
+    const localeFormat = { format: cell.format, locale: this.getters.getLocale() };
     try {
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/cell_evaluation/evaluator.ts:220]: A next step would be to change every throw in this file to returns.

File: src/plugins/ui_core_views/cell_evaluation/evaluator.ts [hunk 7]
```diff
@@ -507,7 +530,7 @@ class PositionBitsEncoder {
   private decodeSheet(sheetKey: PositionId): UID {
     const sheetId = this.inverseSheetMapping.get(sheetKey);
     if (sheetId === undefined) {
-      throw new Error("Sheet id not found");
+      throw { value: ERR_GENERIC, message: "Sheet id not found" };
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/cell_evaluation/evaluator.ts:533]: This is a real programming error. It should remain an `Error`

File: src/plugins/ui_core_views/evaluation_data_validation.ts [hunk 8]
```diff
@@ -228,9 +226,11 @@ export class EvaluationDataValidationPlugin extends UIPlugin {
         );
 
         const evaluated = this.getters.evaluateFormula(sheetId, translatedFormula);
-        return evaluated && !isMatrix(evaluated) ? evaluated.toString() : "";
+        return evaluated?.toString() || "";
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/evaluation_data_validation.ts:229]: what if it's a matrix? Is it safe to remove the check?

File: src/types/cells.ts [hunk 9]
```diff
@@ -25,6 +24,7 @@ export type Cell = LiteralCell | FormulaCell;
 
 interface EvaluatedCellProperties {
   readonly format?: Format;
+  readonly message?: string;
```
Comment (LucasLefevre) [inline src/types/cells.ts:27]: I don't think we should have `message` here since it's specific to a cell value and we do have a specific type for it.
```suggestion
```

File: src/plugins/ui_core_views/cell_evaluation/compilation_parameters.ts [hunk 10]
```diff
@@ -88,29 +88,30 @@ class CompilationParametersBuilder {
...
-      throw new Error(_t("Invalid sheet name"));
+      throw { value: ERR_GENERIC, message: _t("Invalid sheet name: %s") };
     }
     const evaluatedCell = this.getEvaluatedCellIfNotEmpty(position);
     if (evaluatedCell === undefined) {
       return { value: null, format: this.getters.getCell(position)?.format };
     }
-    return evaluatedCell;
+    const { value, format, message } = evaluatedCell;
+    return { value, format, message };
```
Comment (LucasLefevre) [inline src/plugins/ui_core_views/cell_evaluation/compilation_parameters.ts:110]: Let's avoid creating useless objects.
```suggestion
    return evaluatedCell
```

**Review by LucasLefevre — COMMENTED** [review]
Nice !
I think we finally found a good way ! Good job :)

Some general comment (more detailed in individual comments)
- if we find a more meaningful name for `UnitData`, it would be nice
- throwing objects (and not Errors) makes me feel :raised_eyebrow: but :shrug: 
- I think there's some useless changes that could be reverted.
- Now that the functional side is working with this PR, a next step is to improve performance by using returns as much as possible (especially for `ODOO.*` functions)

File: src/formulas/code_builder.ts [hunk 11]
```diff
@@ -3,10 +3,6 @@
  */
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/formulas/code_builder.ts:3]: fixed

File: src/functions/index.ts [hunk 12]
```diff
@@ -74,7 +75,7 @@ class FunctionRegistry extends Registry<FunctionDescription> {
     }
     const descr = addMetaInfoFromArg(addDescr);
     validateArguments(descr.args);
-    this.mapping[name] = createComputeFunctionFromDescription(descr);
+    this.mapping[name] = addErrorHandling(createComputeFunctionFromDescription(descr), name);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/index.ts]: Indeed, to avoid that,  we can put the tryCatch directly in `buildComputeFunctionFromDescription`. @LucasLefevre are you agree with that ?

File: src/functions/helpers.ts [hunk 13]
```diff
@@ -9,7 +9,7 @@ const SORT_TYPES_ORDER = ["number", "string", "boolean", "undefined"];
 
 export function assert(condition: () => boolean, message: string): void {
   if (!condition()) {
-    throw new EvaluationError(CellErrorType.GenericError, message);
+    throw new EvaluationError(message, CellErrorType.GenericError);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/helpers.ts]: That’s what we’re trying to get towards. The "asserts" are intended to be replaced by "returns" with string errors

File: src/helpers/cells/cell_evaluation.ts [hunk 14]
```diff
@@ -60,31 +56,28 @@ export function createEvaluatedCell(
...
 
-function _createEvaluatedCell(value: CellValue | null, localeFormat: LocaleFormat): EvaluatedCell {
-  try {
-    for (const builder of builders) {
-      const evaluateCell = builder(value, localeFormat);
-      if (evaluateCell) {
-        return evaluateCell;
-      }
+function _createEvaluatedCell(value: CellValue, localeFormat: LocaleFormat): EvaluatedCell {
+  if (value instanceof EvaluationError) {
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/helpers/cells/cell_evaluation.ts]: I find it easier to read, less code to have in mind.
Actually, if it really bothers you, I'll go back.

File: tests/evaluation/evaluation.test.ts (hunk 5 above)
Comment (laa-odoo) [reply tests/evaluation/evaluation.test.ts]: Indeed

File: src/helpers/cells/cell_evaluation.ts [hunk 15]
```diff
@@ -155,46 +160,13 @@ function booleanCell(value: boolean, localeFormat: LocaleFormat): BooleanCell {
...
+    value,
+    message,
     isAutoSummable: false,
     defaultAlign: "center",
-    formattedValue: error.errorType,
+    formattedValue: value,
   };
 }
-
-const builders: EvaluatedCellBuilder[] = [
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/helpers/cells/cell_evaluation.ts:169]: we go from 51 lines to 15,
OK the previous code is elegant, it also gives the possibility of easily extending the code to the creation of new cell types.
I may be wrong but I don't think we're considering it.
Here, with 15 lines, in one vision, we understand all the possibilities.
But really I don't care, if you really want to go back then I will

File: src/plugins/ui_core_views/evaluation_data_validation.ts (hunk 8 above)
Comment (laa-odoo) [reply src/plugins/ui_core_views/evaluation_data_validation.ts:229]: I don't know what we want exactly:

Here I removed the check because the try catch below takes care of this case.
The `evaluateFormula` getter has been modified to throw when an array is received.
Perhaps we should first ask ourselves what exactly we want to do with this getters when it returns an array ?

File: src/functions/helpers.ts [hunk 16]
```diff
@@ -233,6 +230,9 @@ export function visitNumbers(args: ArgValue[], cb: (arg: number) => void, locale
       if (typeof cellValue === "number") {
         cb(cellValue);
       }
+      if (cellValue instanceof Error) {
+        throw cellValue;
+      }
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/helpers.ts]: my bad, but it doesn't matter much, it is still managed correctly after

File: src/functions/module_array.ts [hunk 17]
```diff
@@ -584,6 +584,23 @@ const TO_COL_ROW_ARGS = [
   ),
 ];
 
+function dataToKeep(ignore: number): (data: ValueAndFormat) => boolean {
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_array.ts]: ok thx :)

File: src/functions/module_array.ts [hunk 18]
```diff
@@ -594,20 +611,12 @@ export const TOCOL = {
     scanByColumn: Maybe<ValueAndFormat> = { value: TO_COL_ROW_DEFAULT_SCAN }
   ) {
     const _array = toMatrix(array);
-    const _ignore = toInteger(ignore.value, this.locale);
+    const _ignore = toNumber(ignore.value, this.locale);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_array.ts:591]: Done in the `shouldKeepValue` helper

File: src/functions/module_array.ts [hunk 19]
```diff
@@ -594,20 +611,12 @@ export const TOCOL = {
...
     const _scanByColumn = toBoolean(scanByColumn.value);
 
-    assert(() => _ignore >= 0 && _ignore <= 3, _t("Argument ignore must be between 0 and 3"));
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_array.ts:601]: indeed, but this will change again in a commit to turn all "assert" into "return"

File: src/functions/module_statistical.ts [hunk 20]
```diff
@@ -66,6 +66,12 @@ function filterAndFlatData(
       flatDataY.push(valueY);
       flatDataX.push(valueX);
     }
+    if (valueY instanceof Error) {
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_statistical.ts]: Currently all `visitAny` seem to throw when there are errors so yes, 
but let's hope we never need to use it on ranges with errors

File: src/functions/helpers.ts [hunk 21]
```diff
@@ -73,41 +95,37 @@ export function tryToNumber(
...
-export function tryCastAsNumberMatrix(data: Matrix<any>, argName: string): Matrix<number> {
-  data.forEach((row) => {
-    row.forEach((cell) => {
-      if (typeof cell !== "number") {
-        throw new Error(
-          _t(
+export function toNumberMatrix(data: Matrix<UnitData>, argName: string): Matrix<number> {
+  return data.map((row) => {
+    return row.map((cell) => {
+      if (typeof cell.value !== "number") {
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/helpers.ts:102]: `toNumber` will cast values (specially string values), 
and generally we don't want cast values when they come from matrix args

File: src/functions/module_filter.ts [hunk 22]
```diff
@@ -133,19 +133,19 @@ export const FILTER = {
...
       _t("The arguments conditions must have the same dimensions."),
-      ..._conditionsMatrices
+      ...conditions
     );
     const _conditions = _conditionsMatrices.map((c) => c.flat());
 
     const mode = _conditionsMatrices[0].length === 1 ? "row" : "col";
     _array = mode === "row" ? transposeMatrix(_array) : _array;
-
+    conditions;
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_filter.ts:142]: thx

File: src/types/functions.ts [hunk 23]
```diff
@@ -20,18 +19,19 @@ export type ArgType =
...
   type: ArgType[];
   default?: boolean;
   defaultValue?: any;
 }
 
-export type ComputeFunctionArg<T> = T | (() => T);
-export type ComputeFunction<T, R> = (this: EvalContext, ...args: ComputeFunctionArg<T>[]) => R;
+// export type ComputeFunctionArg<T> = T | (() => T);
+// export type ComputeFunction<T, R> = (this: EvalContext, ...args: ComputeFunctionArg<T>[]) => R;
+export type ComputeFunction<T, R> = (this: EvalContext, ...args: T[]) => R;
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/types/functions.ts:32]: indeed

File: src/functions/module_array.ts [hunk 24]
```diff
@@ -340,22 +326,16 @@ export const MINVERSE = {
...
       _matrix
     );
-    if (!isNumberMatrix(_matrix)) {
-      throw new Error(_t("The argument square_matrix must be a matrix of numbers."));
-    }
-
     const { inverted } = invertMatrix(_matrix);
     if (!inverted) {
-      throw new Error(_t("The matrix is not invertible."));
+      throw { value: ERR_GENERIC, message: _t("The matrix is not invertible.") };
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/module_array.ts:338]: Here, all the `throw` in `compute` are cases where `compute` is typed as returning a matrix

File: src/plugins/ui_core_views/cell_evaluation/evaluator.ts (hunk 7 above)
Comment (laa-odoo) [reply src/plugins/ui_core_views/cell_evaluation/evaluator.ts:559]: yes, my bad

File: src/types/misc.ts [hunk 25]
```diff
@@ -181,13 +179,11 @@ export interface RangeCompiledFormula extends Omit<CompiledFormula, "dependencie
 }
 
 export type Matrix<T = unknown> = T[][];
-export type ValueAndFormat = { value: CellValue; format?: Format };
+export type UnitData = { value: CellValue; format?: Format; message?: string };
```
[the remark this answers is not in the document]
Comment (pro-odoo) [reply src/types/misc.ts:182]: `FunctionResult`

File: src/functions/index.ts [hunk 26]
```diff
@@ -74,81 +73,60 @@ class FunctionRegistry extends Registry<FunctionDescription> {
...
-    );
-  }
+function addErrorHandling(
+  compute: ComputeFunction<Arg, Matrix<UnitData> | UnitData>,
+  functionName: string
+): ComputeFunction<Arg, Matrix<UnitData> | UnitData> {
+  return function (this: EvalContext, ...args: Arg[]): Matrix<UnitData> | UnitData {
+    try {
+      const computeFormula = compute.bind(this);
+      return computeFormula(...args);
```
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/functions/index.ts:90]: it is an ideal that we will try to achieve

File: src/types/misc.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (laa-odoo) [reply src/types/misc.ts:182]: `FPayload`

File: src/plugins/ui_core_views/evaluation_data_validation.ts (hunk 8 above)
Comment (LucasLefevre) [reply src/plugins/ui_core_views/evaluation_data_validation.ts:229]: Ok understood. After live discussion: `this.getters.evaluateFormula` shouldn't throw

File: src/formulas/compiler.ts [hunk 27]
```diff
@@ -64,10 +64,10 @@ export function compileTokens(tokens: Token[]): CompiledFormula {
     const scope = new Scope();
 
     if (ast.type === "BIN_OPERATION" && ast.value === ":") {
-      throw new BadExpressionError(_t("Invalid formula"));
+      throw { value: CellErrorType.BadExpression, message: _t("Invalid formula") };
```
Comment (anhe-odoo) [inline src/formulas/compiler.ts:67]: Personnally I prefere the old way to throw/return error. Why did you change this ?

Comment (laa-odoo) [reply src/formulas/compiler.ts:67]: the goal would be that the formulas end up returning errors in the form of the normal payload rather than throwing errors, (performance reason)

Comment (LucasLefevre) [reply src/formulas/compiler.ts:67]: I must admit I agree with Anthony. It feels very here where we are not in the context of a function implementation that would return a value (or an error).

Comment (laa-odoo) [reply src/formulas/compiler.ts:67]: `Bad Expression Error(message)` and other errors could indeed result in `{value: "#SomeError", message}`

**Review by LucasLefevre — COMMENTED** [review]
Yééé nice :)
I think we can merge tomorrow :)
I only left only minor comments (most of them translations)
I did not check commit by commit

**Comment by LucasLefevre** [comment]
robodoo rebase-merge r+

**Comment by LucasLefevre** [comment]
et maintenant ?
robodoo rebase-merge r+

**Comment by LucasLefevre** [comment]
robodoo r+
